src/ImGui.NET.FNA/ImGuiRenderer.cs

```diff
@@ -0,0 +1,313 @@
+
+        private IntPtr? _fontTextureId;
+
+        // Input
+        private int _scrollWheelValue;
+
+        private List<int> _keys = new List<int>();
+
+        public ImGuiRenderer(GraphicsDevice graphicsDevice)
+        {
+            _graphicsDevice = graphicsDevice ?? throw new ArgumentNullException(nameof(graphicsDevice));
+
+            _loadedTextures = new ConcurrentDictionary<IntPtr, Texture2D>();
+
+            _rasterizerState = new RasterizerState()
+            {
+                CullMode = CullMode.None,
+                DepthBias = 0,
+                FillMode = FillMode.Solid,
+                MultiSampleAntiAlias = false,
+                ScissorTestEnable = true,
+                SlopeScaleDepthBias = 0
+            };
+
+            SetupInput();
+        }
+
+        #region ImGuiRenderer
+
+        /// <summary>
+        /// Creates a texture and loads the font data from ImGui. Should be called when the <see cref="GraphicsDevice" /> is initialized but before any rendering is done
+        /// </summary>
+        public virtual void RebuildFontAtlas()
+        {
+            // Get font texture from ImGui
+            var io = ImGui.GetIO();
+            var texData = io.FontAtlas.GetTexDataAsRGBA32();
+
+            // Copy the data to a managed array
+            var pixels = new byte[texData.Width * texData.Height * texData.BytesPerPixel];
+            unsafe { Marshal.Copy(new IntPtr(texData.Pixels), pixels, 0, pixels.Length); }
+
+            // Create and register the texture as an XNA texture
+            var tex2d = new Texture2D(_graphicsDevice, texData.Width, texData.Height, false, SurfaceFormat.Color);
+            tex2d.SetData(pixels);
+
+            // Should a texture already have been build previously, unbind it first so it can be deallocated
+            if (_fontTextureId.HasValue) UnbindTexture(_fontTextureId.Value);
+
+            // Bind the new texture to an ImGui-friendly id
+            _fontTextureId = BindTexture(tex2d);
+
+            // Let ImGui know where to find the texture
+            io.FontAtlas.SetTexID(_fontTextureId.Value);
+            io.FontAtlas.ClearTexData(); // Clears CPU side texture data
+        }
+
+        /// <summary>
+        /// Creates a pointer to a texture, which can be passed through ImGui calls such as <see cref="ImGui.Image" />. That pointer is then used by ImGui to let us know what texture to draw
+        /// </summary>
+        public virtual IntPtr BindTexture(object texture)
+        {
+            // Since this is an XNA-specific renderer, we only support the native Texture2D
+            var tex2d = texture as Texture2D;
+            if (tex2d == null) throw new InvalidOperationException($"Only textures of type '{nameof(Texture2D)}' are supported");
+
+            // The creation of a new texture id and adding it to the local dictionary should be an atomic operation, so we use a lock to ensure that (the concurrent dictionary alone isn't enough)
```

Comment:
Hmm, not saying it hurts to be safe, but most everything else in ImGui is not thread-safe. It's probably unnecessary to worry about it in this particular spot.